### PR TITLE
ENH: Add dtype parameter to IntervalIndex constructors and deprecate from_intervals

### LINKED ISSUE #19262: IntervalIndex constructor doesn't use dtype parameter
#### Code Sample, a copy-pastable example if possible
The `IntervalIndex` constructor takes a `dtype` parameter but doesn't use it (no references to it in the code):

```python
In [2]:  pd.IntervalIndex([pd.Interval(0, 1), pd.Interval(2, 3)], dtype='float64')
Out[2]:
IntervalIndex([(0, 1], (2, 3]]
              closed='right',
              dtype='interval[int64]')
```

The constructor methods do not support a `dtype` parameter:
```python
In [3]: pd.IntervalIndex.from_intervals([pd.Interval(0, 1), pd.Interval(2, 3)], dtype='float64')
---------------------------------------------------------------------------
TypeError: from_intervals() got an unexpected keyword argument 'dtype'

In [4]: pd.IntervalIndex.from_arrays([0, 2], [1, 3], dtype='float64')
---------------------------------------------------------------------------
TypeError: from_arrays() got an unexpected keyword argument 'dtype'

In [5]: pd.IntervalIndex.from_breaks([0, 1, 2, 3], dtype='float64')
---------------------------------------------------------------------------
TypeError: from_breaks() got an unexpected keyword argument 'dtype'

In [6]: pd.IntervalIndex.from_tuples([(0, 1), (2, 3)], dtype='float64')
---------------------------------------------------------------------------
TypeError: from_tuples() got an unexpected keyword argument 'dtype'
```

#### Expected Output

I'd expect the `IntervalIndex` constructor and constructor methods to support a `dtype` parameter, which can be used to override the inferred dtype.

### LINKED ISSUE #19263: Deprecate IntervalIndex.from_intervals in favor of the IntervalIndex constructor
Currently, the `IntervalIndex` constructor and `IntervalIndex.from_intervals` essentially have the same functionality:
```python
In [2]: ivs = [pd.Interval(0, 1), pd.Interval(2, 3)]

In [3]: pd.IntervalIndex(ivs)
Out[3]:
IntervalIndex([(0, 1], (2, 3]]
              closed='right',
              dtype='interval[int64]')

In [4]: pd.IntervalIndex.from_intervals(ivs)
Out[4]:
IntervalIndex([(0, 1], (2, 3]]
              closed='right',
              dtype='interval[int64]')
```

However, both are using independent code to do this, which seems redundant.  This has also led to two being slightly out of sync, as the `IntervalIndex` constructor supports a few more parameters (`dtype`, `closed`) and has some additional checks that `IntervalIndex.from_intervals` does not.

Seems like there are a few options to clean things up:
  1. Move all code to the constructor, and redirect `from_intervals` to the constructor
  2. Move all construction code to `from_intervals`, and redirect the constructor to `from_intervals`
      - Would still have some constructor specific behavior, e.g. `fastpath`
  3. Remove `from_intervals` entirely, and only use the constructor
  4. Something else?

I'm leaning towards 1, as it looks like that's what eventually happens elsewhere for other `from_*` methods, e.g. `Categorical.from_codes`, `MultiIndex.from_product`.  Don't really have a strong opinion though.

### PR DESCRIPTION
**Pull request by jschendel**
- [X] closes #19262
- [X] closes #19263
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

Summary:
- Added support for a `dtype` parameter to all IntervalIndex constructors
   - Allows users to override the inferred dtype
- Deprecated `IntervalIndex.from_intervals`
   - Still added `dtype` parameter, since it's just a pass through to `IntervalIndex`
   - Removed usage and references to `IntervalIndex.from_intervals` throughout the codebase 
- Split construction tests off into `interval/test_construction.py`
   - Created a base class for common tests, and subclasses for each constructor
   - Was previously written in a more flat style, where each constructor was explicitly called
   - Expanded the tests to hit some previously untested behavior

### CODE REVIEW

File: doc/source/api.rst [hunk 1]
```diff
@@ -1617,7 +1617,6 @@ IntervalIndex Components
    IntervalIndex.from_arrays
    IntervalIndex.from_tuples
    IntervalIndex.from_breaks
-   IntervalIndex.from_intervals
```
Comment (jschendel) [inline doc/source/api.rst]: Should this be removed from api.rst now, or should I wait until `from_intervals` is fully removed and not just deprecated?

Comment (jreback) [reply doc/source/api.rst]: yes this is fine

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -200,7 +200,8 @@ def __new__(cls, data=None, dtype=None, copy=False, name=None,
         # interval
         if is_interval_dtype(data) or is_interval_dtype(dtype):
             from .interval import IntervalIndex
-            return IntervalIndex(data, dtype=dtype, name=name, copy=copy)
+            return IntervalIndex(data, dtype=dtype, name=name, copy=copy,
+                                 **kwargs)
```
Comment (jreback) [inline pandas/core/indexes/base.py]: do we actually need kwargs to be passed or is that for consistentcy with others?

File: pandas/core/indexes/interval.py [hunk 3]
```diff
@@ -151,6 +152,8 @@ class IntervalIndex(IntervalMixin, Index):
         Name to be stored in the index.
     copy : boolean, default False
         Copy the meta-data
+    dtype : dtype or None, default None
+        If None, dtype will be inferred
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: add a version added

File: pandas/core/indexes/interval.py [hunk 4]
```diff
@@ -417,6 +427,8 @@ def from_breaks(cls, breaks, closed='right', name=None, copy=False):
             Name to be stored in the index.
         copy : boolean, default False
             copy the data
+        dtype : dtype or None, default None
+            If None, dtype will be inferred
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: vesionadded

File: pandas/core/indexes/interval.py [hunk 5]
```diff
@@ -458,6 +469,8 @@ def from_arrays(cls, left, right, closed='right', name=None, copy=False):
...
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: same

File: pandas/core/indexes/interval.py [hunk 6]
```diff
@@ -545,6 +559,8 @@ def from_tuples(cls, data, closed='right', name=None, copy=False):
             Name to be stored in the index.
         copy : boolean, default False
             by-default copy the data, this is compat only and ignored
+        dtype : dtype or None, default None
+            If None, dtype will be inferred
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: same

File: pandas/core/indexes/interval.py [hunk 7]
```diff
@@ -571,15 +585,25 @@ def from_tuples(cls, data, closed='right', name=None, copy=False):
             if isna(d):
                 lhs = rhs = np.nan
             else:
-                lhs, rhs = d
+                try:
+                    lhs, rhs = d
+                except ValueError:
+                    msg = ('IntervalIndex.from_tuples requires tuples of '
+                           'length 2, got {tpl}').format(tpl=d)
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: are we testing both of these paths?

File: pandas/core/indexes/interval.py [hunk 8]
```diff
@@ -571,15 +585,25 @@ def from_tuples(cls, data, closed='right', name=None, copy=False):
...
+                except TypeError:
+                    msg = ('IntervalIndex.from_tuples received an invalid '
+                           'item, {tpl}').format(tpl=d)
+                    raise TypeError(msg)
             left.append(lhs)
             right.append(rhs)
 
         # TODO
         # if we have nulls and we previous had *only*
         # integer data, then we have changed the dtype
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: I think this comment is moot? (this was from me originally).....

**Review by jreback — CHANGES_REQUESTED** [review]
small comments. testing looks great!

File: pandas/core/indexes/base.py (hunk 2 above)
Comment (jschendel) [reply pandas/core/indexes/base.py]: It's needed to pass `closed`.

File: pandas/core/indexes/interval.py (hunk 3 above)
Comment (jschendel) [reply pandas/core/indexes/interval.py]: done

File: pandas/core/indexes/interval.py (hunk 4 above)
Comment (jschendel) [reply pandas/core/indexes/interval.py]: done

File: pandas/core/indexes/interval.py (hunk 5 above)
Comment (jschendel) [reply pandas/core/indexes/interval.py]: done

File: pandas/core/indexes/interval.py (hunk 6 above)
Comment (jschendel) [reply pandas/core/indexes/interval.py]: done

File: pandas/core/indexes/base.py (hunk 2 above)
Comment (jreback) [reply pandas/core/indexes/base.py]: can you pass that directly?

File: pandas/core/indexes/interval.py [hunk 9]
```diff
@@ -245,19 +247,28 @@ def __new__(cls, data, closed=None,
...
+                    dtype=None, verify_integrity=True):
         result = IntervalMixin.__new__(cls)
 
-        if closed is None:
-            closed = 'right'
+        closed = closed or 'right'
         left = _ensure_index(left, copy=copy)
         right = _ensure_index(right, copy=copy)
 
+        if dtype is not None:
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: can you add a 1-liner here, noting that this must be an interval dtype

File: pandas/core/indexes/interval.py [hunk 10]
```diff
@@ -571,15 +595,21 @@ def from_tuples(cls, data, closed='right', name=None, copy=False):
             if isna(d):
                 lhs = rhs = np.nan
             else:
-                lhs, rhs = d
+                try:
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: can you add a 1-liner here showing the format of th expected input (list-of-tuples)

**Review by jreback — APPROVED** [review]
very minor comments. ping on green.

File: pandas/core/indexes/base.py (hunk 2 above)
Comment (jschendel) [reply pandas/core/indexes/base.py]: Done by doing `closed = kwargs.get('closed', None)`, then passing that into the constructor.  Not sure if you meant doing it that way, or by adding `closed=None` to the `Index` constructor.

File: pandas/core/indexes/interval.py (hunk 9 above)
Comment (jschendel) [reply pandas/core/indexes/interval.py]: done

File: pandas/core/indexes/interval.py (hunk 10 above)
Comment (jschendel) [reply pandas/core/indexes/interval.py]: done

**Comment by jschendel** [comment]
@jreback : green

**Comment by jreback** [comment]
nice @jschendel keep em coming!

**Pull request merged**